Firestore/core/src/firebase/firestore/local/query_cache.h

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_LOCAL_QUERY_CACHE_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_LOCAL_QUERY_CACHE_H_
+
+#if !defined(__OBJC__)
+#error "For now, this file must only be included by ObjC source files."
+#endif  // !defined(__OBJC__)
+
+#import <Foundation/Foundation.h>
+
+#include "Firestore/core/src/firebase/firestore/model/document_key.h"
+#include "Firestore/core/src/firebase/firestore/model/document_key_set.h"
+#include "Firestore/core/src/firebase/firestore/model/snapshot_version.h"
+#include "Firestore/core/src/firebase/firestore/model/types.h"
+
+@class FSTQuery;
+@class FSTQueryData;
+
+NS_ASSUME_NONNULL_BEGIN
+
+namespace firebase {
+namespace firestore {
+namespace local {
+
+/**
+ * Represents cached targets received from the remote backend. This contains
+ * both a mapping between targets and the documents that matched them according
+ * to the server, but also metadata about the targets.
+ *
+ * The cache is keyed by FSTQuery and entries in the cache are FSTQueryData
+ * instances.
+ */
+class QueryCache {
+ public:
+  typedef void (^TargetEnumerator)(FSTQueryData*, BOOL*);
```

Comment:
This is an OK intermediate step, but to get fully to C++ we shouldn't be using an Objective-C block for this. That's not available when compiling with something other than clang (e.g. with MSVC).

I think you want something like `std::function<void(FSTQueryData*)>`, or if you need the early exit use `bool` as the return type?

Also, prefer `using` to `typedef`, so `using TargetEnumerator = std::function<void(FSTQueryData*)>;`.

Reply:
I agree with all of that, but I think the right time to do it is when the LRU delegates are ported, since they are the callers and are currently in ObjC and pass around accept and pass blocks.